flink-runtime/src/test/java/org/apache/flink/runtime/state/IncrementalRemoteKeyedStateHandleTest.java

```diff
@@ -243,6 +244,60 @@ public void testNonEmptyIntersection() {
         assertEquals(handle.getStateHandleId(), newHandle.getStateHandleId());
     }
 
+    @Test
+    public void testConcurrentCheckpointSharedStateRegistration() throws Exception {
+        StateHandleID handleID = new StateHandleID("1.sst");
+        StreamStateHandle streamHandle1 = new ByteStreamStateHandle("file-1", new byte[] {'s'});
+        StreamStateHandle streamHandle2 = new ByteStreamStateHandle("file-2", new byte[] {'s'});
+
+        SharedStateRegistry registry = new SharedStateRegistryImpl();
+
+        UUID backendID = UUID.randomUUID();
+
+        IncrementalRemoteKeyedStateHandle handle1 =
+                new IncrementalRemoteKeyedStateHandle(
+                        backendID,
+                        KeyGroupRange.of(0, 0),
+                        1L,
+                        placeSpies(
+                                new HashMap<StateHandleID, StreamStateHandle>() {
+                                    {
+                                        put(handleID, streamHandle1);
+                                    }
+                                }),
+                        Collections.emptyMap(),
+                        new ByteStreamStateHandle("", new byte[] {'s'}));
+
+        handle1.registerSharedStates(registry, handle1.getCheckpointId());
+
+        IncrementalRemoteKeyedStateHandle handle2 =
+                new IncrementalRemoteKeyedStateHandle(
+                        backendID,
+                        KeyGroupRange.of(0, 0),
+                        2L,
+                        placeSpies(
+                                new HashMap<StateHandleID, StreamStateHandle>() {
+                                    {
+                                        put(handleID, streamHandle2);
+                                    }
+                                }),
+                        Collections.emptyMap(),
+                        new ByteStreamStateHandle("", new byte[] {'s'}));
+
+        handle2.registerSharedStates(registry, handle2.getCheckpointId());
+
+        registry.checkpointCompleted(1L);
+        registry.checkpointCompleted(2L);
```

Comment:
Shouldn't this 2nd checkpoint be aborted? Otherwise, why would any state be discarded?